### PR TITLE
Improved country selector

### PR DESCRIPTION
**Pull request by hamideha**
### PR Summary: 

**Designer responsible:** @designrichly 
**Developer responsible:**: @hamideha 

Improves the country selectors in Dawn.

1. Adds basic search functionality
2. Shows store's popular countries at the top of the list. This uses `popular?` on the Country object in liquid which is true for at most 4 countries on a shop.
3. Improves mobile responsiveness of country selector. On mobile instead of having a dropdown, the selector is essentially a modal dialog.
4. Some UI improvements
   - Move the country to show up before the currency
   - Use a checkmark instead of underline to indicate that a country is selected

### Why are these changes introduced?

As part of the mission of increasing cross-border conversion, we decided to improve the first piece of the online store that a cross-border buyer interacts with.

User-testing was done with a prototype of the new selectors and they outperformed the current selectors. The users used the popular countries, they often searched for countries, and they commented on how the current selectors are less user-friendly than the prototype. 

### What approach did you take?

The code that was added, for the most part, follows patterns already established in Dawn. Nothing unique or out of the ordinary. Possibly the only thing that is different is the search functionality (it's more of a filter than search); it uses a very naive and basic filtering approach with string matching (`.includes(string)`) 

The no-js approach to country selectors is exactly the same as it is today; using a browser-native select element.

### Decision log

Some UX decisions that were made:
| # | Decision | Alternatives | Rationale | Downsides |
|---|---|---|---|---|
| 1 | By default we only popular countries if there are more than 9 total countries and if there is more than 1 popular country |  | If there are less than 9 countries, the popular country would appear very close to its identical copy in the regular list |   |
| 2 | We only show the search box if there are more than 9 countries |  | The list would be short enough that users don't have to search |   |

### Visual impact on existing themes
1. New search box in the country selector
2. New mobile design
3. Small UI nits

### Testing steps/scenarios
Just play around with the country selector and the search functionality. There are 3 places that the country selector could show up: the header, the footer, and the announcement bar (only on desktop).

### Demo links
Store without popular countries:
- [Store](https://hamidehs-store.myshopify.com/)
- [Editor](https://admin.shopify.com/store/hamidehs-store/themes/162597994518?key=snippets%2Fcountry-localization.liquid)

Store with popular countries (Note that the popular countries are hardcoded):
- [Store](https://mylittleearthling.myshopify.com/)
- [Editor](https://potteryplants.myshopify.com/admin/themes/137609019605?key=snippets%2Fcountry-localization.liquid)

All 1st party themes:
[Sense](https://hamidehs-store.myshopify.com/?_ab=0&_fd=0&_sc=1&preview_theme_id=162694168598)
[Crave](https://hamidehs-store.myshopify.com/?_ab=0&_fd=0&_sc=1&preview_theme_id=162694201366)
[Craft](https://hamidehs-store.myshopify.com/?_ab=0&_fd=0&_sc=1&preview_theme_id=162694266902)
[Studio](https://hamidehs-store.myshopify.com/?_ab=0&_fd=0&_sc=1&preview_theme_id=162694299670)
[Taste](https://hamidehs-store.myshopify.com/?_ab=0&_fd=0&_sc=1&preview_theme_id=162694332438)
[Colorblock](https://hamidehs-store.myshopify.com/?_ab=0&_fd=0&_sc=1&preview_theme_id=162694365206)
[Ride](https://hamidehs-store.myshopify.com/?_ab=0&_fd=0&_sc=1&preview_theme_id=162694397974)
[Refresh](https://hamidehs-store.myshopify.com/?_ab=0&_fd=0&_sc=1&preview_theme_id=162694430742)
[Origin](https://hamidehs-store.myshopify.com/?_ab=0&_fd=0&_sc=1&preview_theme_id=162694463510)
[Publisher](https://hamidehs-store.myshopify.com/?_ab=0&_fd=0&_sc=1&preview_theme_id=162694496278)
[Spotlight](https://hamidehs-store.myshopify.com/?_ab=0&_fd=0&_sc=1&preview_theme_id=162694529046)

### Checklist
- [x] Requested review from UX (Only for changes that are affecting the experience or perceivable visual details)
- [x] Followed [theme code principles](https://github.com/Shopify/dawn/blob/main/.github/CONTRIBUTING.md#theme-code-principles)
- [x] Linted with [Theme Check](https://github.com/Shopify/theme-check)
- [x] Tested on [mobile](https://shopify.dev/themes/store/requirements#mobile-browser-requirements)
- [x] Tested on [multiple browsers](https://shopify.dev/themes/store/requirements#desktop-browser-requirements)
- [x] Tested for [accessibility](https://shopify.dev/themes/best-practices/accessibility)

### CODE REVIEW

**Comment by designrichly** [comment]
I'll hold off my review until I've synced with @kjellr , but one thing I'd like to test is seeing the search icon on the unfocused state.

**Comment by hamideha** [comment]
I have signed the CLA!

File: sections/footer.liquid
Comment (hamideha) [reply sections/footer.liquid:251]: @Roi-Arthur thanks for the suggestion! I explored your method and the issue is that the country selector would still be bound by the height of the footer. A screen with this resolution is very unlikely, but in the image below I set the height so that more content other than the footer is visible and setting the overlay's height to 100vh doesn't actually take 100% of the viewport but rather 100% of the footer's height.

Screenshot 2023-12-11 at 12 53 01 PM

**Comment by hamideha** [comment]
@ludoboludo @eugenekasimov @andrewetchen Is there any way to expedite the review on this PR? I'm really sorry for the quick notice but we'd like to merge this by the Editions cutoff and I head off for the holidays on Wednesday.

File: sections/footer.liquid
Comment (Roi-Arthur) [reply sections/footer.liquid:251]: Uh, interesting 🤔
In my example the footer didn't take the whole screen (you can see the Subscribe panel but also the end of the section above, with the Shop all button.

In order to preserve the animation for desktop, would you then instead consider changing the CSS?
Currently [Dawn has this on line 3289](https://github.com/hamideha/dawn/blob/hamideha/selector/assets/base.css#L3289)
```css
.scroll-trigger.animate--slide-in {
    transform: translateY(2rem);
  }
```

We could add something to disable that transform in the footer and on mobile? Using the !important doesn't feel great but is necessary because the animation css kicks in and overrides it otherwise 😞
```css
.scroll-trigger.animate--slide-in {
    transform: translateY(2rem);
  }

@media screen and (max-width: 749px) {
  .footer__content-bottom.scroll-trigger.animate--slide-in {
    transform: none !important;
  }
}
```

Comment (hamideha) [reply sections/footer.liquid]: Ah yes that works and is a cleaner solution. Thank you!

Comment (ludoboludo) [reply sections/footer.liquid:251]: So am I understanding correctly the comments and the animation class should be re added ?

File: assets/component-localization-form.css
Comment (ludoboludo) [inline assets/component-localization-form.css:334]: This is working for the country selection but not language. We should be adding the same icon to the language picker otherwise it's hard to tell what is currently selected: 
<details>
<summary>Screenshot</summary>


</details>

File: assets/component-menu-drawer.css
Comment (ludoboludo) [inline assets/component-menu-drawer.css:273]: I don't think this was meant to be applied for the menu drawer country/language selectors but for the footer one 🤔

File: snippets/country-localization.liquid
Comment (ludoboludo) [inline snippets/country-localization.liquid:80]: I wonder if there would be some value in having a clear label as to what this list is compared to the second one. When I use a screen reader, I could see it being useful to understand the difference 🤔

File: snippets/country-localization.liquid
Comment (ludoboludo) [inline snippets/country-localization.liquid:77]: Something that I notice as well is that when I open the country selector, it reads all the current list elements in there. 
So I open it, then it starts telling me all the currencies 😅 
The `aria-live="polite"` seem to be the issue. I think we could have it worked differently. Right now if I type as well something in the search, it's reading out all the list to me... I wonder if that is useful really to a user. Maybe if it's just a hidden div we update with a text saying `list updated` that isn't as invasive as mentioning all the items in the list. 

Not too sure, what's your take on it ?

File: assets/component-menu-drawer.css
Comment (hamideha) [reply assets/component-menu-drawer.css:273]: One of the UX feedbacks we received regarding the country selector in the menu drawer is that we should also overlay the header so that we don't see 2 close buttons, since clicking the menu drawer's close button would also close the country selector and this might be a source of confusion. These styles allow us to do that. 

Screenshot 2023-12-14 at 8 45 30 AM

File: assets/component-localization-form.css
Comment (hamideha) [reply assets/component-localization-form.css:334]: Good catch. Thank you!

File: sections/footer.liquid
Comment (hamideha) [reply sections/footer.liquid:251]: I haven't added them back because testing on other WebKit browsers, specifically on iOS versions lower than 17 
```css
.footer__content-bottom.scroll-trigger.animate--slide-in {
    transform: none !important;
}
```
doesn't work. I'm still trying to figure out so I haven't added them back yet.

**Review by eugenekasimov — COMMENTED** [review]
In Safari the desktop selectors in the header and footer look different and not elegant in my opinion, especially in the footer because the search is located on the top and you cannot see it until you scroll all the way up.
https://screenshot.click/14-46-m19ev-tw4pe.mp4

File: snippets/country-localization.liquid
Comment (hamideha) [reply snippets/country-localization.liquid:77]: @ludoboludo which screen reader are you using? Using Mac's voice over I can't reproduce it reading all the countries and currencies in the list when typing.

Comment (hamideha) [reply snippets/country-localization.liquid:77]: I agree, though, that it's not a friendly experience to have all the countries and currencies read out and will change it.

Comment (ludoboludo) [reply snippets/country-localization.liquid:77]: Yeah it was with Voice Over on Safari. Were you doing voice over and chrome ? 🤔 Ill try again. 
Here is [a recording](https://screenshot.click/15-13-l3qaq-hpewi.mp4)

**Comment by eugenekasimov** [comment]
I found one more small issue that only occurs in Safari.

When you open the country selector, whether in the header or footer, and start scrolling, you may notice that the list of countries extends slightly beyond the top edge of the main container. 
https://screenshot.click/22-52-ekjde-uhvkj.mp4

File: snippets/country-localization.liquid
Comment (ludoboludo) [inline snippets/country-localization.liquid:39]: I think a class name that would be a "modifier" following BEM convention would be better and less generic. 

```suggestion
    <div class="country-selector__actions {% unless show_country_filter %}country-selector__actions--no-padding{% endunless %}">
```

It's definitely longer but at least not as generic. If it was a class we use in different spots for the same outcome it would make sense but for now I think it's only used here.

File: assets/section-footer.css
Comment (ludoboludo) [inline assets/section-footer.css:76]: Why are we removing the animation on the mobile footer again?

File: snippets/language-localization.liquid
Comment (ludoboludo) [inline snippets/language-localization.liquid:33]: Nitpick. So we don't output an empty class attribute. 
```suggestion
            <span {% if language.iso_code != localization.language.iso_code %}class="visibility-hidden"{% endif %}>
```

File: assets/component-localization-form.css
Comment (hamideha) [reply assets/component-localization-form.css:151]: Ahh thanks for pointing this out! The solution here was to simply use the `keydown` event instead of `keyup` for the up and down arrows.

**Comment by hamideha** [comment]
@ludoboludo PR is ready for review again. I've addressed your comments but let me know if there's anything else that should be fixed.

File: assets/component-localization-form.css
Comment (hamideha) [reply assets/component-localization-form.css:179]: @ludoboludo both these issues should be fixed now.

File: assets/localization-form.js
Comment (ludoboludo) [inline assets/localization-form.js:47]: You will probably need to update this one as well to make sure it's added when needed and not on tablet sizes.

File: assets/base.css
Comment (ludoboludo) [inline assets/base.css:3133]: Here I feel like we don't really need any of those `country-filter__input` added to those CSS declarations since it's already getting it from `field__input` no ? Or am I missing something 🤔

**Review by ludoboludo — COMMENTED** [review]
Looks like we might need to check in with the team that works on the translation platform. As we need to make sure the translations are back in before merging the PR.

File: assets/base.css
Comment (hamideha) [reply assets/base.css:3133]: Yep you're 100% right. Initially I wasn't using `field__input` and I just never thought of removing these once I did start using them.

File: assets/localization-form.js
Comment (hamideha) [reply assets/localization-form.js:110]: The way I ended up fixing this is by only adding `overflow-hidden-mobile` if `overflow-hidden-tablet` isn't already a class on the body.

File: assets/localization-form.js
Comment (hamideha) [reply assets/localization-form.js:47]: This line removes the class so I think it's fine as is

**Review by tyleralsbury — COMMENTED** [review]
There are two things that I noticed.  I wouldn't say that either are blockers, but I would at least create issues to follow up on them.

---

When you have sticky header open and open up the dropdown in the footer on mobile, the header is over the background.  It can still be clicked to close the modal, but if you click the logo it will go to the homepage.  Ideally clicking out of the modal would just close it.

Screenshot 2024-01-04 at 8 58 53 AM

---

Secondly, because of its similarity in behaviour to the predictive search bar, I believe we should add support for arrow keys to go through the results.  For an example, make a search in the predictive search bar and then use your up and down arrow keys.  You'll see a highlight and can hit enter to go to the corresponding page.

This is nice for UX and for accessibility.

**Review by ludoboludo — APPROVED** [review]
I think we're looking good 👍 
Thanks for addressing the feedback

**Pull request closed without merging**
